### PR TITLE
build: bake npm dependencies into builder image

### LINKED ISSUE #2058: ci: bake npm and related ui dependencies into builder image
Every CI build reinstalls mocha, stylus, tslint, typescript, chai and probably some other items. It should be possible bake these deps into the builder image. See [build/builder.sh init](https://github.com/cockroachdb/cockroach/blob/master/build/builder.sh#L7). The old `deps-docker.sh` script accomplished this for the `cockroachdb/cockroach-dev` image, but I think the technique will have to be modified for `build/builder.sh init`. Or maybe not, I haven't looked in depth.

### PR DESCRIPTION
**Pull request by tamird**
Closes #2058.

### CODE REVIEW

**Comment by tamird** [comment]
This doesn't work. I think we'll need to change `ui/Makefile` to use globally-installed npm stuff in CI. Ideally it would still automatically `npm install` (note absence of `-g`) when said dependencies are not otherwise in the PATH, but I'm not sure how (or if it's even possible) to express this in Make. Anyone know this magic? @bdarnell @petermattis

File: build/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+ apt-get install --no-install-recommends --auto-remove -y git build-essential file npm nodejs && \
+ apt-get clean autoclean && \
+ apt-get autoremove -y && \
+ rm -rf /tmp/* && \
+ ln -s /usr/bin/nodejs /usr/bin/node
+RUN go get golang.org/x/tools/cmd/vet
+
+COPY npm-shrinkwrap.json package.json ./
+RUN npm install
+RUN rm npm-shrinkwrap.json package.json
```
Comment (petermattis) [inline build/Dockerfile]: Each action creates a new docker layer which is why you often see these long command lines separated by `&&`. Probably worthwhile to do:

```
  RUN npm install && \
    rm npm-shrinkwrap.json package.json
```

Might even be worthwhile to combine with the `RUN` above.

**Comment by petermattis** [comment]
I don't quite understand what `ui/Makefile` is doing. Looks like the dependencies are all installed to `ui/node_modules` which we can't do when `cockroachdb/builder` is created because we don't have a source tree. Perhaps `circle-deps.sh` can install a symlink from wherever `cockroachdb/builder` is installing `node_modules` to the source tree and also touch `npm.installed`.

File: build/Dockerfile (hunk 1 above)
Comment (tamird) [inline build/Dockerfile]: Yeah, I combined it with all but the apt-get `RUN`; I wouldn't want `npm` changes to invalidate that cache.

**Comment by tamird** [comment]
OK, so this works now. It saves about 2 seconds, which doesn't seem worth the extra complexity.

**Comment by petermattis** [comment]
How did you measure the improvement? The rationale for doing this was that sometimes the npm installation takes minutes, presumably because of networking issues.

**Comment by tamird** [comment]
I looked at the output of `time ${builder} /bin/bash -c "(go generate ...` which was 19s without this and 17s with.

**Comment by petermattis** [comment]
Right. Usually the `go generate` step is fast, but sometimes it is very very slow due to the npm installation.

**Comment by tamird** [comment]
Is it possible that circle CI added a caching npm proxy since #2058 was filed? `npm install` takes 8 seconds on my machine, FWIW.

**Comment by petermattis** [comment]
I see no evidence for a caching npm proxy. My guess is that npm installation is usually fast and sometimes very very slow and the culprit could be a whole host of places.

**Pull request closed without merging**